File: runtime/typescript/src/result.ts
import { DecodeCursor, Decoder, EncodeCursor, Encoder } from './index';

export namespace Result {
  export class Ok<Ok, Err> {
    ok: Ok;
    constructor(ok: Ok) {
      this.ok = ok;
    }
    toString(): string {
      const stringified = this.ok.toString();
      if (stringified == "[object Object]") {
        return `Ok(${JSON.stringify(this.ok)})`;
      } else {
        return `Ok(${stringified})`;
      }
    }
  }
  export class Err<Ok, Err> {
    err: Err;
    constructor(err: Err) {
      this.err = err;
    }
    toString(): string {
      const stringified = this.err.toString();
      if (stringified == "[object Object]") {
        return `Err(${JSON.stringify(this.err)})`;
      } else {
        return `Err(${stringified})`;
      }
    }
  }
}

export type Result<Ok, Err> = Result.Ok<Ok, Err> | Result.Err<Ok, Err>;

export class ResultEncoder<Ok, Err> implements Encoder<Result<Ok, Err>>, Decoder<Result<Ok, Err>> {
  private okEncoder: Encoder<Ok> & Decoder<Ok>;
  private errEncoder: Encoder<Err> & Decoder<Err>;

  constructor(
    okEncoder: Encoder<Ok> & Decoder<Ok>,
    errEncoder: Encoder<Err> & Decoder<Err>,
  ) {
    this.okEncoder = okEncoder;
    this.errEncoder = errEncoder;
  }

  baseLength = () => 1 + Math.max(this.okEncoder.baseLength(), this.errEncoder.baseLength());

  scratchLength(value: Result<Ok, Err>): number {
    if (value instanceof Result.Ok) {
      return this.okEncoder.scratchLength(value.ok);
    } else if (value instanceof Result.Err) {
      return this.errEncoder.scratchLength(value.err);
    } else {
      throw "Failed to encode Result - value is not a Result";
    }
  }

  encode(cursor: EncodeCursor, value: Result<Ok, Err>) {
    if (value instanceof Result.Ok) {
      cursor.buffer.setUint8(cursor.base(1), 0);
      this.okEncoder.encode(cursor, value.ok);
    } else if (value instanceof Result.Err) {
      cursor.buffer.setUint8(cursor.base(1), 1);
      this.errEncoder.encode(cursor, value.err);
    } else {
      throw "Failed to encode Result - value is not a Result";
    }
  }

  decode(cursor: DecodeCursor): Result<Ok, Err> {
    let variant = cursor.buffer.getUint8(cursor.base(1));

    if (variant == 0) {
      return new Result.Ok(this.okEncoder.decode(cursor));
    } else if (variant == 1) {
      return new Result.Err(this.errEncoder.decode(cursor));
    } else {
      throw "Failed to decode Result - invalid variant tag";
    }
  }
}

export class ResultLazyDecoder<Ok, Err> implements Decoder<Result<Ok, Err>> {
  private okDecoder: Decoder<Ok>;
  private errDecoder: Decoder<Err>;

  constructor(
    okDecoder: Decoder<Ok>,
    errDecoder: Decoder<Err>,
  ) {
    this.okDecoder = okDecoder;
    this.errDecoder = errDecoder;
  }

  baseLength = () => 1 + Math.max(this.okDecoder.baseLength(), this.errDecoder.baseLength());

  decode(cursor: DecodeCursor): Result<Ok, Err> {
    let variant = cursor.buffer.getUint8(cursor.base(1));

    if (variant == 0) {
      return new Result.Ok(this.okDecoder.decode(cursor));
    } else if (variant == 1) {
      return new Result.Err(this.errDecoder.decode(cursor));
    } else {
      throw "Failed to decode Result - invalid variant tag";
    }
  }
}

export const ProtoResult = <Ok, Err>(okEncoder: Encoder<Ok> & Decoder<Ok>, errEncoder: Encoder<Err> & Decoder<Err>) => new ResultEncoder(okEncoder, errEncoder);
export const ProtoResultLazy = <Ok, Err>(okDecoder: Decoder<Ok>, errDecoder: Decoder<Err>) => new ResultLazyDecoder(okDecoder, errDecoder);

File: runtime/rust/src/result.rs
use crate::{
    max, BaseLen, Compatible, Decode, DecodeCursor, DecodeError, DecodeResult, Encode,
    EncodeCursor, Lazy, Owned,
};

impl<O: Owned, E: Owned> Owned for Result<O, E> {
    type Lazy<'a> = Result<O::Lazy<'a>, E::Lazy<'a>>;

    fn lazy_to_owned(lazy: Self::Lazy<'_>) -> DecodeResult<Self> {
        match lazy {
            Ok(ok) => Ok(Ok(O::lazy_to_owned(ok)?)),
            Err(err) => Ok(Err(E::lazy_to_owned(err)?)),
        }
    }
}

impl<'a, O: Lazy<'a>, E: Lazy<'a>> Lazy<'a> for Result<O, E> {
    type Owned = Result<O::Owned, E::Owned>;
}

impl<T: BaseLen, E: BaseLen> BaseLen for Result<T, E> {
    const BASE_LEN: usize = 1 + max(T::BASE_LEN, E::BASE_LEN);
}

impl<T: Encode, E: Encode> Encode for Result<T, E> {
    fn scratch_len(&self) -> usize {
        match self {
            Ok(ok) => ok.scratch_len(),
            Err(err) => err.scratch_len(),
        }
    }

    fn encode(&self, cursor: &mut EncodeCursor) {
        match self {
            Ok(ok) => {
                cursor.base(1)[0] = 0;
                ok.encode(cursor);
                cursor.base(Self::BASE_LEN - 1 - T::BASE_LEN);
            }
            Err(err) => {
                cursor.base(1)[0] = 1;
                err.encode(cursor);
                cursor.base(Self::BASE_LEN - 1 - E::BASE_LEN);
            }
        }
    }
}

impl<'a, T: Decode<'a>, E: Decode<'a>> Decode<'a> for Result<T, E> {
    fn decode(cursor: &DecodeCursor<'a>) -> DecodeResult<Self> {
        let variant = cursor.base(1)[0];
        if variant > 1 {
            return Err(DecodeError);
        }

        let is_ok = variant == 0;
        if is_ok {
            let ok = T::decode(cursor)?;
            cursor.advance(Self::BASE_LEN - 1 - T::BASE_LEN);
            Ok(Ok(ok))
        } else {
            let err = E::decode(cursor)?;
            cursor.advance(Self::BASE_LEN - 1 - E::BASE_LEN);
            Ok(Err(err))
        }
    }
}

impl<Ok1, Err1, Ok2, Err2> Compatible<Result<Ok2, Err2>> for Result<Ok1, Err1>
where
    Ok1: Compatible<Ok2>,
    Err1: Compatible<Err2>,
    Ok2: Encode,
    Err2: Encode,
{
}
